language-plutus-core/plutus-ir/Language/PlutusIR/TypeCheck/Internal.hs

```diff
@@ -0,0 +1,615 @@
+      -- Checks that tycons+tyargs are unique among eachother, and dataconstructors are unique among eachother
+      -- TODO: move these 2 checks to a different pass, e.g. in renaming phase
+      assertNoDuplicate allNewTyNames
+      assertNoDuplicate allNewNames
+      -- check the data-constructors' to be kinded by *
+      for_ vdecls $ \(VarDecl annV _ ty) -> do
+                  -- we normalize the type, since the user might have written some type-computation in the dataconstructor.
+                  -- we split the normalized type to a bunch of constructors arguments and a result-type.
+                  Normalized normTy <- normalizeTypeM ty
+                  let actualArgsTypes = funTyArgs normTy
+                      actualResType = funResultType normTy
+                  -- Check that the result-type is *-kinded, inside the whole env scope.
+                  -- The result-type can see all tyvardecls (tyargs + currently-defined type constructor).
+                  withTyVarDecls allNewTyDecls $
+                      checkKindM annV actualResType $ Type ()
+                  -- Check the dataconstructor arguments' types to also be *-kinded,
+                  -- but the environment defers for rec and nonrec bindings.
+                  withTyVarDecls
+                      (case recurs of
+                           -- With the whole env  in scope.
+                           Rec -> allNewTyDecls
+                           -- The type-constructor is *not visible* in the arguments' types.
+                           NonRec -> tyargs
+                      ) $ -- check each argument in the defined environment.
+                          for_ actualArgsTypes $ \actualArgType ->
+                             checkKindM annV actualArgType $ Type ()
+                  -- Check for wellformedness of the result-type to be in the expected form (see `expectedDataConsResType`)
+                  -- Note1: in the paper this check for the result-type welformnedness is not needed,
+                  -- because the result-type is "implicit" (and not explicit in the FIR syntax).
+                  when (expectedDataConsResType /= void actualResType) $
+                      throwError $ MalformedDataConstrResType annV expectedDataConsResType
+   where
+      allNewTyDecls :: [TyVarDecl TyName ann]
+      allNewTyDecls = tvdecl:tyargs
+
+      allNewTyNames :: [T.Text]
+      allNewTyNames = fmap (nameString . unTyName . tyVarDeclName) allNewTyDecls
+
+      allNewNames :: [T.Text]
+      allNewNames = nameString des : fmap (nameString . varDeclName) vdecls
+
+      assertNoDuplicate :: [T.Text] -> TypeCheckM uni ann ()
+      assertNoDuplicate = maybe (pure ())
+                                (throwError . DuplicateDeclaredIdent annD)
+                          . hasDuplicate
+
+      withTyVarDecls :: [TyVarDecl TyName ann] -> TypeCheckM uni ann () -> TypeCheckM uni ann ()
```

Comment:
I think it would make your life easier to just define `withVarDecl`, `withTyVarDecl`, and their plural versions at the top level. e.g. there are a couple of places where you deconstruct decls to use `withVar` etc and I think it would be nicer to just have versions that worked on the decls.